File: java/com/tigervnc/rfb/Configuration.java
/* Copyright (C) 2002-2005 RealVNC Ltd.  All Rights Reserved.
 * Copyright 2004-2005 Cendio AB.
 * Copyright 2012 Brian P. Hinz
 *
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this software; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,
 * USA.
 */

//
// Configuration - class for dealing with configuration parameters.
//

package com.tigervnc.rfb;

import java.io.FileInputStream;
import java.io.PrintWriter;
import java.lang.reflect.Field;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.*;

import com.tigervnc.vncviewer.VncViewer;

public class Configuration {

  static LogWriter vlog = new LogWriter("Configuration");

  private static final String IDENTIFIER_STRING = "TigerVNC Configuration file Version 1.0";

  public enum ConfigurationObject { ConfGlobal, ConfServer, ConfViewer };

  // -=- The Global/server/viewer Configuration objects
  private static Configuration global_ = null;
  private static Configuration server_ = null;
  private static Configuration viewer_ = null;

  public static Configuration global() {
    if (global_ == null)
      global_ = new Configuration("Global");
    return global_;
  }

  public static Configuration server() {
    if (server_ == null)
      server_ = new Configuration("Server");
    return server_;
  }

  public static Configuration viewer() {
    if (viewer_ == null)
      viewer_ = new Configuration("Viewer");
    return viewer_;
  }

  // Enable server/viewer specific parameters
  public static void enableServerParams() { global().appendConfiguration(server()); }
  public static void enableViewerParams() { global().appendConfiguration(viewer()); }

  // Append configuration object to this instance.
  // NOTE: conf instance can be only one configuration object
  public void appendConfiguration(Configuration conf) {
    conf._next = _next; _next = conf;
  }

  // -=- Configuration implementation
  public Configuration(String name_, Configuration attachToGroup) {
    name = name_; head = null; _next = null;
    if (attachToGroup != null) {
      _next = attachToGroup._next;
      attachToGroup._next = this;
    }
  }

  public Configuration(String name_) {
    this(name_, null);
  }

  // - Return the buffer containing the Configuration's name
  final public String getName() { return name; }

  // - Assignment operator.  For every Parameter in this Configuration's
  //   group, get()s the corresponding source parameter and copies its
  //   content.
  public Configuration assign(Configuration src) {
    VoidParameter current = head;
    while (current != null) {
      VoidParameter srcParam = ((Configuration)src).get(current.getName());
      if (srcParam != null) {
        current.immutable = false;
        String value = srcParam.getValueStr();
        vlog.debug("operator=("+current.getName()+", "+value+")");
        current.setParam(value);
      }
      current = current._next;
    }
    if (_next != null)
      _next = src;
    return this;
  }

  // - Set named parameter to value
  public boolean set(String n, String v, boolean immutable) {
    return set(n, n.length(), v, immutable);
  }

  public boolean set(String n, String v) {
    return set(n, n.length(), v, false);
  }

  // - Set parameter to value (separated by "=")
  public boolean set(String name, int len,
                     String val, boolean immutable)
  {
    VoidParameter current = head;
    while (current != null) {
      if (current.getName().length() == len &&
          current.getName().equalsIgnoreCase(name.substring(0, len)))
      {
        boolean b = current.setParam(val);
        current.setHasBeenSet();
        if (b && immutable)
  	  current.setImmutable();
        return b;
      }
      current = current._next;
    }
    return (_next != null) ? _next.set(name, len, val, immutable) : false;
  }

  // - Set named parameter to value, with name truncated at len
  boolean set(String config, boolean immutable) {
    boolean hyphen = false;
    if (config.charAt(0) == '-') {
      hyphen = true;
      config = config.substring(1);
      if (config.charAt(0) == '-') config = config.substring(1); // allow gnu-style --<option>
    }
    int equal = config.indexOf('=');
    if (equal > -1) {
      return set(config, equal, config.substring(equal+1), immutable);
    } else if (hyphen) {
      VoidParameter current = head;
      while (current != null) {
        if (current.getName().equalsIgnoreCase(config)) {
          boolean b = current.setParam();
  	  current.setHasBeenSet();
          if (b && immutable)
  	    current.setImmutable();
          return b;
        }
        current = current._next;
      }
    }
    return (_next != null) ? _next.set(config, immutable) : false;
  }

  boolean set(String config) {
    return set(config, false);
  }

  // - Container for process-wide Global parameters
  public static boolean setParam(String param, String value, boolean immutable) {
    return global().set(param, value, immutable);
  }

  public static boolean setParam(String param, String value) {
    return setParam(param, value, false);
  }

  public static boolean setParam(String config, boolean immutable) {
    return global().set(config, immutable);
  }

  public static boolean setParam(String config) {
    return setParam(config, false);
  }

  public static boolean setParam(String name, int len,
                                 String val, boolean immutable) {
    return global().set(name, len, val, immutable);
  }


  // - Get named parameter
  public VoidParameter get(String param)
  {
    VoidParameter current = head;
    while (current != null) {
      if (current.getName().equalsIgnoreCase(param))
        return current;
      current = current._next;
    }
    return (_next != null) ? _next.get(param) : null;
  }

  public static VoidParameter getParam(String param) { return global().get(param); }

  public static void listParams(int width, int nameWidth) {
    global().list(width, nameWidth);
  }
  public static void listParams() {
    listParams(79, 10);
  }

  public void list(int width, int nameWidth) {
    VoidParameter current = head;

    System.err.format("%s Parameters:%n", name);
    while (current != null) {
      String def_str = current.getDefaultStr();
      String desc = current.getDescription().trim();
      String format = "  %-"+nameWidth+"s -";
      System.err.format(format, current.getName());
      int column = current.getName().length();
      if (column < nameWidth) column = nameWidth;
      column += 4;
      while (true) {
        int s = desc.indexOf(' ');
        int wordLen;
        if (s > -1) wordLen = s;
        else wordLen = desc.length();

        if (column + wordLen + 1 > width) {
          format = "%n%"+(nameWidth+4)+"s";
          System.err.format(format, "");
          column = nameWidth+4;
        }
        format = " %"+wordLen+"s";
        System.err.format(format, desc.substring(0, wordLen));
        column += wordLen + 1;

        if (s == -1) break;
        desc = desc.substring(wordLen+1);
      }

      if (def_str != null) {
        if (column + def_str.length() + 11 > width)
          System.err.format("%n%"+(nameWidth+4)+"s","");
        System.err.format(" (default=%s)%n",def_str);
        def_str = null;
      } else {
        System.err.format("%n");
      }
      current = current._next;
    }

    if (_next != null)
      _next.list(width, nameWidth);
  }

  public void list() {
    list(79, 10);
  }

  // Name for this Configuration
  private String name;

  // - Pointer to first Parameter in this group
  public VoidParameter head;

  // Pointer to next Configuration in this group
  public Configuration _next;

}

File: rfb/Configuration.cxx
/* Copyright (C) 2002-2004 RealVNC Ltd.  All Rights Reserved.
 * Copyright (C) 2004-2005 Cendio AB. All rights reserved.
 *    
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this software; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
 * USA.
 */

// -=- Configuration.cxx

#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <assert.h>
#ifdef WIN32
#define strcasecmp _stricmp
#define strncasecmp _strnicmp
#endif

#include <rfb/util.h>
#include <rfb/Configuration.h>
#include <rfb/LogWriter.h>
#include <rfb/Exception.h>

#ifdef WIN32

// Under Win32, we use these routines from several threads,
// so we must provide suitable locking.
#include <rfb/Threading.h>

static rfb::Mutex configLock;

#endif

#include <rdr/HexOutStream.h>
#include <rdr/HexInStream.h>

using namespace rfb;

static LogWriter vlog("Config");


// -=- Configuration

VoidParameter* Configuration::head = 0;

bool Configuration::setParam(const char* n, const char* v, bool immutable) {
  return setParam(n, strlen(n), v, immutable);
}

bool Configuration::setParam(const char* name, int len,
                             const char* val, bool immutable)
{
  VoidParameter* current = head;
  while (current) {
    if ((int)strlen(current->getName()) == len &&
        strncasecmp(current->getName(), name, len) == 0)
    {
      bool b = current->setParam(val);
      current->setHasBeenSet(); 
      if (b && immutable) 
	current->setImmutable();
      return b;
    }
    current = current->_next;
  }
  return false;
}

bool Configuration::setParam(const char* config, bool immutable) {
  bool hyphen = false;
  if (config[0] == '-') {
    hyphen = true;
    config++;
    if (config[0] == '-') config++; // allow gnu-style --<option>
  }
  const char* equal = strchr(config, '=');
  if (equal) {
    return setParam(config, equal-config, equal+1, immutable);
  } else if (hyphen) {
    VoidParameter* current = head;
    while (current) {
      if (strcasecmp(current->getName(), config) == 0) {
        bool b = current->setParam();
	current->setHasBeenSet(); 
        if (b && immutable) 
	  current->setImmutable();
        return b;
      }
      current = current->_next;
    }
  }    
  return false;
}

VoidParameter* Configuration::getParam(const char* param)
{
  VoidParameter* current = head;
  while (current) {
    if (strcasecmp(current->getName(), param) == 0)
      return current;
    current = current->_next;
  }
  return 0;
}

void Configuration::listParams(int width, int nameWidth) {
  VoidParameter* current = head;
  while (current) {
    char* def_str = current->getDefaultStr();
    const char* desc = current->getDescription();
    fprintf(stderr,"  %-*s -", nameWidth, current->getName());
    int column = strlen(current->getName());
    if (column < nameWidth) column = nameWidth;
    column += 4;
    while (true) {
      const char* s = strchr(desc, ' ');
      int wordLen;
      if (s) wordLen = s-desc;
      else wordLen = strlen(desc);

      if (column + wordLen + 1 > width) {
        fprintf(stderr,"\n%*s",nameWidth+4,"");
        column = nameWidth+4;
      }
      fprintf(stderr," %.*s",wordLen,desc);
      column += wordLen + 1;
      desc += wordLen + 1;
      if (!s) break;
    }

    if (def_str) {
      if (column + (int)strlen(def_str) + 11 > width)
        fprintf(stderr,"\n%*s",nameWidth+4,"");
      fprintf(stderr," (default=%s)\n",def_str);
      strFree(def_str);
    } else {
      fprintf(stderr,"\n");
    }
    current = current->_next;
  }
}

// -=- VoidParameter

VoidParameter::VoidParameter(const char* name_, const char* desc_)
  : immutable(false), _hasBeenSet(false), name(name_), description(desc_) {
  _next = Configuration::head;
  Configuration::head = this;
}

VoidParameter::~VoidParameter() {
}

const char*
VoidParameter::getName() const {
  return name;
}

const char*
VoidParameter::getDescription() const {
  return description;
}

bool VoidParameter::setParam() {
  return false;
}

bool VoidParameter::isBool() const {
  return false;
}

void
VoidParameter::setImmutable() {
  vlog.debug("set immutable %s", getName());
  immutable = true;
}

void
VoidParameter::setHasBeenSet() {
  _hasBeenSet = true;
} 

bool
VoidParameter::hasBeenSet() {
  return _hasBeenSet;
}

// -=- AliasParameter

AliasParameter::AliasParameter(const char* name_, const char* desc_,
                               VoidParameter* param_)
  : VoidParameter(name_, desc_), param(param_) {
}

bool
AliasParameter::setParam(const char* v) {
  return param->setParam(v);
}

bool AliasParameter::setParam() {
  return param->setParam();
}

char*
AliasParameter::getDefaultStr() const {
  return 0;
}

char* AliasParameter::getValueStr() const {
  return param->getValueStr();
}

bool AliasParameter::isBool() const {
  return param->isBool();
}

void
AliasParameter::setImmutable() {
  vlog.debug("set immutable %s (Alias)", getName());
  param->setImmutable();
}


// -=- BoolParameter

BoolParameter::BoolParameter(const char* name_, const char* desc_, bool v)
: VoidParameter(name_, desc_), value(v), def_value(v) {
}

bool
BoolParameter::setParam(const char* v) {
  if (immutable) return true;

  if (*v == 0 || strcasecmp(v, "1") == 0 || strcasecmp(v, "on") == 0
      || strcasecmp(v, "true") == 0 || strcasecmp(v, "yes") == 0)
    value = 1;
  else if (strcasecmp(v, "0") == 0 || strcasecmp(v, "off") == 0
           || strcasecmp(v, "false") == 0 || strcasecmp(v, "no") == 0)
    value = 0;
  else {
    vlog.error("Bool parameter %s: invalid value '%s'", getName(), v);
    return false;
  }

  vlog.debug("set %s(Bool) to %s(%d)", getName(), v, value);
  return true;
}

bool BoolParameter::setParam() {
  setParam(true);
  return true;
}

void BoolParameter::setParam(bool b) {
  if (immutable) return;
  value = b;
  vlog.debug("set %s(Bool) to %d", getName(), value);
}

char*
BoolParameter::getDefaultStr() const {
  char* result = new char[8];
  sprintf(result, "%d", (int)def_value);
  return result;
}

char* BoolParameter::getValueStr() const {
  char* result = new char[8];
  sprintf(result, "%d", (int)value);
  return result;
}

bool BoolParameter::isBool() const {
  return true;
}

BoolParameter::operator bool() const {
  return value;
}

// -=- IntParameter

IntParameter::IntParameter(const char* name_, const char* desc_, int v)
: VoidParameter(name_, desc_), value(v), def_value(v) {
}

bool
IntParameter::setParam(const char* v) {
  if (immutable) return true;
  vlog.debug("set %s(Int) to %s", getName(), v);
  value = atoi(v);
  return true;
}

bool
IntParameter::setParam(int v) {
  if (immutable) return true;
  vlog.debug("set %s(Int) to %d", getName(), v);
  value = v;
  return true;
}

char*
IntParameter::getDefaultStr() const {
  char* result = new char[16];
  sprintf(result, "%d", def_value);
  return result;
}

char* IntParameter::getValueStr() const {
  char* result = new char[16];
  sprintf(result, "%d", value);
  return result;
}

IntParameter::operator int() const {
  return value;
}

// -=- StringParameter

StringParameter::StringParameter(const char* name_, const char* desc_,
                                 const char* v)
  : VoidParameter(name_, desc_), value(strDup(v)), def_value(v)
{
  if (!v) {
    fprintf(stderr,"Default value <null> for %s not allowed\n",name_);
    throw rfb::Exception("Default value <null> not allowed");
  }
}

StringParameter::~StringParameter() {
  strFree(value);
}

bool StringParameter::setParam(const char* v) {
#ifdef WIN32
  Lock l(configLock);
#endif
  if (immutable) return true;
  if (!v)
    throw rfb::Exception("setParam(<null>) not allowed");
  vlog.debug("set %s(String) to %s", getName(), v);
  strFree(value);
  value = strDup(v);
  return value != 0;
}

char* StringParameter::getDefaultStr() const {
  return strDup(def_value);
}

char* StringParameter::getValueStr() const {
#ifdef WIN32
  Lock l(configLock);
#endif
  return strDup(value);
}

// -=- BinaryParameter

BinaryParameter::BinaryParameter(const char* name_, const char* desc_, const void* v, int l)
: VoidParameter(name_, desc_), value(0), length(0), def_value((char*)v), def_length(l) {
  if (l) {
    value = new char[l];
    length = l;
    memcpy(value, v, l);
  }
}
BinaryParameter::~BinaryParameter() {
  if (value)
    delete [] value;
}

bool BinaryParameter::setParam(const char* v) {
#ifdef WIN32
  Lock l(configLock);
#endif
  if (immutable) return true;
  vlog.debug("set %s(Binary) to %s", getName(), v);
  return rdr::HexInStream::hexStrToBin(v, &value, &length);
}

void BinaryParameter::setParam(const void* v, int len) {
#ifdef WIN32
  Lock l(configLock);
#endif
  if (immutable) return; 
  vlog.debug("set %s(Binary)", getName());
  delete [] value; value = 0;
  if (len) {
    value = new char[len];
    length = len;
    memcpy(value, v, len);
  }
}

char* BinaryParameter::getDefaultStr() const {
  return rdr::HexOutStream::binToHexStr(def_value, def_length);
}

char* BinaryParameter::getValueStr() const {
#ifdef WIN32
  Lock l(configLock);
#endif
  return rdr::HexOutStream::binToHexStr(value, length);
}

void BinaryParameter::getData(void** data_, int* length_) const {
#ifdef WIN32
  Lock l(configLock);
#endif
  if (length_) *length_ = length;
  if (data_) {
    *data_ = new char[length];
    memcpy(*data_, value, length);
  }
}
